hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -488,91 +478,62 @@ void Vgen::emit(const call& i) {
   }
 }
 
-void Vgen::emit(const callm& i) {
-  a->Ldr(rAsm, M(i.target));
-  a->Blr(rAsm);
-}
-
 void Vgen::emit(const calls& i) {
   emitSmashableCall(a->code(), env.meta, i.target);
 }
 
 ///////////////////////////////////////////////////////////////////////////////
 
-void Vgen::emit(const stublogue& i) {
-  // Push FP, LR always regardless of i.saveframe (makes SP 16B aligned)
-  emit(pushp{rfp(), rlr()});
-}
-
-void Vgen::emit(const stubret& i) {
-  if (i.saveframe) {
-    emit(popp{rfp(), rlr()});
-  } else {
-    emit(popp{PhysReg(rAsm), rlr()});
-  }
-  emit(ret{});
-}
-
-void Vgen::emit(const callstub& i) {
-  emit(call{i.target, i.args});
+void Vgen::emit(const callphp& i) {
+  emitSmashableCall(a->code(), env.meta, i.stub);
+  emit(unwind{{i.targets[0], i.targets[1]}});
 }
 
-void Vgen::emit(const callfaststub& i) {
+void Vgen::emit(const callarray& i) {
   emit(call{i.target, i.args});
-  emit(syncpoint{i.fix});
-}
-
-void Vgen::emit(const tailcallstub& i) {
-  // Pop off FP/LR pair and jump to target
-  emit(popp{rfp(), rlr()});
-  emit(jmpi{i.target, i.args});
 }
 
-///////////////////////////////////////////////////////////////////////////////
+void Vgen::emit(const contenter& i) {
+  vixl::Label stub, end;
 
-void Vgen::emit(const phplogue& i) {
-  // Save LR in m_savedRip on the current VM frame pointed by 'i.fp'
-  a->Str(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-}
+  // Jump past the stub below
+  a->B(&end);
 
-void Vgen::emit(const phpret& i) {
-  a->Ldr(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-  if (!i.noframe) {
-    a->Ldr(X(i.d), X(i.fp)[AROFF(m_sfp)]);
-  }
-  emit(ret{});
-}
+  // We call into this stub from the end below. Take that
+  // LR and store it in m_savedRip. Then jump to the target.
+  a->bind(&stub);
+  a->Str(X(rlr()), M(i.fp[AROFF(m_savedRip)]));
+  a->Br(X(i.target));
 
-void Vgen::emit(const callphp& i) {
-  emitSmashableCall(a->code(), env.meta, i.stub);
+  // Call to stub above and then unwind
+  a->bind(&end);
+  a->Bl(&stub);
   emit(unwind{{i.targets[0], i.targets[1]}});
 }
 
-void Vgen::emit(const tailcallphp& i) {
-  // To make callee's return as caller's return, load the return address at
-  // i.fp[AROFF(m_savedRip)] into LR and jmp to target
-  a->Ldr(X(rlr()), X(i.fp)[AROFF(m_savedRip)]);
-  emit(jmpr{i.target, i.args});
-}
+///////////////////////////////////////////////////////////////////////////////
 
-void Vgen::emit(const callarray& i) {
-  emit(call{i.target, i.args});
-}
+void Vgen::emit(const calltc& i) {
+  vixl::Label stub;
 
-void Vgen::emit(const contenter& i) {
-  vixl::Label End;
+  // Preserve the FP and LR (exittc) on the stack.
```

Comment:
May not need to, but we did it since we have to push in pairs. ARM has to push it's LR before doing any branch with link (bl/blr), and so it's typically done with FP in that pair making a new frame. `leavetc` below restores it...again although it might not need to. The previous implementation for `calltc` has the same comment, but it was pretty wrong as it pushed savedRIP along with LR.